warp/src/raygun/mod.rs

```diff
@@ -548,41 +548,54 @@ impl Default for ConversationSettings {
 #[repr(C)]
 pub struct DirectConversationSettings {}
 
-#[derive(Default, Debug, Copy, Clone, Serialize, Deserialize, PartialEq, Eq, Hash, Display)]
-#[display(
-    fmt = "Everyone can add participants: {}, Everyone can change name: {}",
-    "if self.members_can_add_participants {\"✅\"} else {\"❌\"}",
-    "if self.members_can_change_name {\"✅\"} else {\"❌\"}"
-)]
+#[derive(Default, Debug, Clone, Serialize, Deserialize, PartialEq, Eq, Display)]
+#[display(fmt = "Roles: {:?}", "self.roles")]
 #[repr(C)]
 pub struct GroupSettings {
-    // Everyone can add participants, if set to `true``.
-    #[serde(default)]
-    members_can_add_participants: bool,
-    // Everyone can change the name of the group.
+    // Roles to manage permissions. Key is the unique name of the role.
     #[serde(default)]
-    members_can_change_name: bool,
+    roles: IndexMap<String, Role>,
+}
+
+#[derive(Clone, Eq, PartialEq, Serialize, Deserialize, Debug)]
+pub struct Role {
+    // Participants assigned to this role
+    participants: Vec<DID>,
+    // Permissions of this role
+    permissions: Vec<Permission>,
+}
+
+#[derive(Clone, Eq, PartialEq, Serialize, Deserialize, Debug)]
+pub enum Permission {
+    All,
+    SetRoles,
+    ChangeGroupName,
+    AddParticipantsToGroup,
 }
 
 impl GroupSettings {
     pub fn new() -> Self {
         Self::default()
     }
 
-    pub fn members_can_add_participants(&self) -> bool {
-        self.members_can_add_participants
-    }
-
-    pub fn members_can_change_name(&self) -> bool {
-        self.members_can_change_name
+    pub fn roles(&mut self) -> IndexMap<String, Role> {
+        self.roles.clone()
     }
 
-    pub fn set_members_can_add_participants(&mut self, val: bool) {
-        self.members_can_add_participants = val;
+    pub fn set_roles(&mut self, roles: IndexMap<String, Role>) {
+        self.roles = roles;
     }
 
-    pub fn set_members_can_change_name(&mut self, val: bool) {
-        self.members_can_change_name = val;
+    pub fn user_has_permission(&self, id: &DID, permission: Permission) -> bool {
+        for (_, role) in &self.roles {
```

Comment:
```suggestion
        for role in self.roles.values() {
```

Reply:
Although as I think about it, we could probably reduce this to a single call by using `values` as the iterator and using `any` with the check for permission

Reply:
removed roles from group chats